vault/client.go

```diff
@@ -27,58 +31,177 @@ func NewClient(logger *log.Logger, vaultAuthRole, vaultAuthProvider string, vaul
+}
+
+// login authenticates to Vault using IAM auth, and sets the client's token.
+func login(ctx context.Context, stsSvc *sts.STS, client *api.Client, authConfig config.AuthConfig) (*api.Secret, error) {
 	// ignore out
 	req, _ := stsSvc.GetCallerIdentityRequest(&sts.GetCallerIdentityInput{})
+	req.SetContext(ctx)
 
-	if vaultIAMServerID != "" {
-		req.HTTPRequest.Header.Add("X-Vault-AWS-IAM-Server-ID", vaultIAMServerID)
+	if authConfig.IAMServerID != "" {
+		req.HTTPRequest.Header.Add("X-Vault-AWS-IAM-Server-ID", authConfig.IAMServerID)
 	}
 
 	if signErr := req.Sign(); signErr != nil {
-		return nil, nil, signErr
+		return nil, signErr
 	}
 
 	headers, err := json.Marshal(req.HTTPRequest.Header)
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
 	body, err := ioutil.ReadAll(req.HTTPRequest.Body)
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
 	d := make(map[string]interface{})
 	d["iam_http_request_method"] = req.HTTPRequest.Method
 	d["iam_request_url"] = base64.StdEncoding.EncodeToString([]byte(req.HTTPRequest.URL.String()))
 	d["iam_request_headers"] = base64.StdEncoding.EncodeToString(headers)
 	d["iam_request_body"] = base64.StdEncoding.EncodeToString(body)
-	d["role"] = vaultAuthRole
+	d["role"] = authConfig.Role
 
-	logger.Println("attemping Vault login...")
-	resp, err := vaultClient.Logical().Write(fmt.Sprintf("auth/%s/login", vaultAuthProvider), d)
+	secret, err := client.Logical().Write(fmt.Sprintf("auth/%s/login", authConfig.Provider), d)
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
-	if resp == nil {
-		return nil, nil, fmt.Errorf("got no response from the %s authentication provider", vaultAuthProvider)
+	if secret == nil {
+		return nil, fmt.Errorf("got no response from the %s authentication provider", authConfig.Provider)
 	}
 
-	token, err := parseToken(resp)
+	token, err := secret.TokenID()
 	if err != nil {
-		return nil, nil, fmt.Errorf("error parsing token: %s", err)
+		return nil, fmt.Errorf("error reading token: %s", err)
 	}
-	vaultClient.SetToken(token)
+	client.SetToken(token)
 
-	return vaultClient, config, nil
+	return secret, nil
+}
+
+// refreshToken will handle renewing the existing token until it reaches its maximum
+// lease and will continue re-authenticating and renewing until the context is cancelled.
+// Should be called immediately after a token is received.
+func refreshToken(ctx context.Context, logger *log.Logger, stsSvc *sts.STS, client *api.Client, config config.AuthConfig, secret *api.Secret) {
+	for {
+		// Sleep for the first 2/3 of the initial token TTL, as the renewer below
+		// will immediately start with a renew call, and in many cases we expect
+		// the Lambda function will exit before the first renewal is even needed.
+		err := sleepUntilRenewingShouldStart(ctx, secret)
+		if err != nil {
+			logger.Printf("error while waiting for first token renewal: %s\n", err)
+		}
+
+		err = renewToken(ctx, logger, client, secret)
```

Comment:
Does this function block? I think it does, it might be good to document that at the function definition and maybe even call that out here